nv2a-decode.py

```diff
@@ -0,0 +1,753 @@
+
+def decode_in_reg(word, shift, is_alpha = False, is_final_combiner_a = False, is_final_combiner_bcd = False):
+  v = word >> shift
+  mapping = (v >> 5) & 0x7
+  alpha = (v >> 4) & 1
+  source = (v >> 0) & 0xF
+
+  is_final_combiner_abcd = is_final_combiner_a or is_final_combiner_bcd
+  assert(is_final_combiner_bcd or source != 0xE) # SPARE0_PLUS_SECONDARY_COLOR_NV
+  assert(is_final_combiner_abcd or source != 0xF) # E_TIMES_F_NV
+
+  s = input_regs[source]
+
+  #FIXME: This might be wrong?
+  if not is_alpha:
+    s += ".aaa" if alpha else ".rgb"
+  else:
+    assert(source != 3) # FIXME: This is supposed to look for FOG
+    s += ".a" if alpha else ".b"
+  return mappings_code[mapping] % s
+
+def decode_register_combiner_stage(stage, is_alpha):
+
+  if not is_alpha:
+    combine_input = read_word(nv2a_mem, 0x401900 + stage * 4)
+  else:
+    combine_input = read_word(nv2a_mem, 0x4018C0 + stage * 4)
+
+  a = decode_in_reg(combine_input, 24, is_alpha)
+  b = decode_in_reg(combine_input, 16, is_alpha)
+  c = decode_in_reg(combine_input, 8, is_alpha)
+  d = decode_in_reg(combine_input, 0, is_alpha)
+
+  if not is_alpha:
+    combine_output = read_word(nv2a_mem, 0x401920 + stage * 4)
+
+    #FIXME: Support these!
+    b_to_a_ab = (combine_output >> 19) & 1
+    b_to_a_cd = (combine_output >> 18) & 1
+  else:
+    combine_output = read_word(nv2a_mem, 0x4018E0 + stage * 4)
+    b_to_a_ab = False
+    b_to_a_cd = False
+
+  op = (combine_output >> 15) & 7
+  ops = ('NOSHIFT',
+         'NOSHIFT_BIAS',
+         'SHIFTLEFTBY1',
+         'SHIFTLEFTBY1_BIAS',
+         'SHIFTLEFTBY2',
+         'SHIFTRIGHTBY1')
+  ops_code = ('%s',
+              '(%s - 0.5)',
+              '(%s * 2)',
+              '((%s - 0.5) * 2)',
+              '(%s * 4)',
+              '(%s / 2)')
+  mux_enable = (combine_output >> 14) & 1
+  ab_dot_enable = (combine_output >> 13) & 1
+  cd_dot_enable = (combine_output >> 12) & 1
+  sum_dst = (combine_output >> 8) & 0xF
+  ab_dst = (combine_output >> 4) & 0xF
+  cd_dst = (combine_output >> 0) & 0xF
+
+  op_code = ops_code[op]
+
+  def get_out(is_dot_product, gcc_x, gcc_y, gcc_z):
+
+    if not is_alpha:
+
+      if is_dot_product == False:
+        res_str = op_code % gcc_x
+      else:
+        res_str = op_code % gcc_y
+
+    else:
+
+      if is_dot_product == False:
+        res_str = op_code % gcc_z
+      else:
+        res_str = "<invalid>" # FIXME: Assert?
+        assert(False)
+
+    return res_str
+
+  # "gcc" stands for general combiner computation.
+
+  if not is_alpha:
+    gcc1 = "%s * %s" % (a, b)
+    gcc2 = "vec3(dot(%s, %s))" % (a, b)
+    gcc3 = "%s * %s" % (c, d)
+    gcc4 = "vec3(dot(%s, %s))" % (c, d)
+    gcc5 = "%s + %s" % (gcc1, gcc3)
+    gcc6 = "mix(%s, %s, SPARE0_NV.a & %s)" % (gcc3, gcc1, mux_bit_str)
+  else:
+    gcc1 = "%s * %s" % (a, b)
+    gcc2 = "%s * %s" % (c, d)
+    gcc3 = "%s + %s" % (gcc1, gcc2)
+    gcc4 = "mix(%s, %s, SPARE0_NV.a & %s)" % (gcc2, gcc1, mux_bit_str)
```

Comment:
don't resolve these vars here.. just put the logic of the used ones into the code